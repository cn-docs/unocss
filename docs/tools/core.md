---
title: 核心
description: UnoCSS 的核心引擎，不包含任何预设。它可以作为你自己原子 CSS 框架的引擎使用。
---

# 核心

UnoCSS 的核心引擎（无预设）：`@unocss/core`。它可以作为你自己原子 CSS 框架的引擎使用。

## 安装

::: code-group

```bash
pnpm add -D @unocss/core
```

```bash
yarn add -D @unocss/core
```

```bash
npm install -D @unocss/core
```

:::

## 使用

```ts
import { createGenerator } from '@unocss/core'

const generator = await createGenerator(
  { /* 用户选项 */ },
  { /* 默认选项 */ }
)

const { css } = await generator.generate(code)
```

## 许可证

- MIT 许可证 &copy; 2021-PRESENT [Anthony Fu](https://github.com/antfu)
