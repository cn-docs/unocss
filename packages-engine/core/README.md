# @unocss/core

The core engine of [UnoCSS](https://github.com/unocss/unocss) without any presets. It can be used as the engine of your own atomic CSS framework.

## Documentation

Please refer to the [documentation](https://unocss.dev/tools/core).

## License

MIT License &copy; 2021-PRESENT [Anthony Fu](https://github.com/antfu)
