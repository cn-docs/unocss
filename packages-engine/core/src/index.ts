export * from './config'
export * from './constants'
export * from './extractors'
export * from './generator'
export * from './types'
export * from './utils'
