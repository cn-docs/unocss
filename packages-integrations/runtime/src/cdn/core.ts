import init from '../index'

init()
