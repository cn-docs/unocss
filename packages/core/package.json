{
  "name": "@unocss/core",
  "type": "module",
  "version": "0.62.3",
  "description": "The instant on-demand Atomic CSS engine.",
  "author": "Anthony Fu <anthonyfu117@hotmail.com>",
  "license": "MIT",
  "funding": "https://github.com/sponsors/antfu",
  "homepage": "https://github.com/unocss/unocss/tree/main/packages/core#readme",
  "repository": {
    "type": "git",
    "url": "https://github.com/unocss/unocss",
    "directory": "packages/core"
  },
  "bugs": {
    "url": "https://github.com/unocss/unocss/issues"
  },
  "keywords": [
    "unocss",
    "atomic-css",
    "atomic-css-engine",
    "css",
    "tailwind",
    "windicss"
  ],
  "sideEffects": false,
  "exports": {
    ".": {
      "types": "./dist/index.d.mts",
      "default": "./dist/index.mjs"
    }
  },
  "main": "dist/index.mjs",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "unbuild",
    "stub": "unbuild --stub"
  },
  "devDependencies": {
    "magic-string": "catalog:",
    "unconfig": "catalog:"
  }
}
