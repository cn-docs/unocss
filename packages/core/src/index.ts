export * from './config'
export * from './extractors'
export * from './generator'
export * from './types'
export * from './utils'
