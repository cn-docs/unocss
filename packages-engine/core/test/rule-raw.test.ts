import { createGenerator, e } from '@unocss/core'
import { describe, expect, it } from 'vitest'

describe('rule-raw', () => {
  it('basic', async () => {
    const uno = await createGenerator({
      presets: [],
      rules: [
        ['before', { content: 'before' }],
        [/^m-(.+)$/, ([, d], { rawSelector }) => {
          if (d.includes('hi')) {
            return `
/* return a string to take full control of the generated CSS */
/* this also means you have to take care of all the things yourself */
/* for example, escaping, applying variants, etc. */
.${e(rawSelector)} {
  border-bottom: 1px solid green;
}
.${e(rawSelector)}::after {
  content: ${JSON.stringify(d)};
  color: skyblue;
}
.hi ${e(rawSelector)}::before {
  content: 'hi';
}
`.trim()
          }
        }],
        // the other two rules are making sure the sorting is correct
        [/^m-(\d+)$/, ([, d]) => ({ margin: `${d}px` })],
      ],
    })
    const { css } = await uno.generate('m-1 m-hi-anthony m-hi-hi before', { preflights: false })
    expect(css).toMatchSnapshot('')
  })
})
