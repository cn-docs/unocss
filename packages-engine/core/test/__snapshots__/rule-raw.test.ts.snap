// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rule-raw > basic 1`] = `
"/* layer: default */
.before{content:before;}
/* return a string to take full control of the generated CSS */
/* this also means you have to take care of all the things yourself */
/* for example, escaping, applying variants, etc. */
.m-hi-anthony {
  border-bottom: 1px solid green;
}
.m-hi-anthony::after {
  content: "hi-anthony";
  color: skyblue;
}
.hi m-hi-anthony::before {
  content: 'hi';
}
/* return a string to take full control of the generated CSS */
/* this also means you have to take care of all the things yourself */
/* for example, escaping, applying variants, etc. */
.m-hi-hi {
  border-bottom: 1px solid green;
}
.m-hi-hi::after {
  content: "hi-hi";
  color: skyblue;
}
.hi m-hi-hi::before {
  content: 'hi';
}
.m-1{margin:1px;}"
`;
